### DISCUSSION TITLE
Bug when returning an object from an action in Next.js

### DISCUSSION THREAD

**Discussion by samuelgomez4** [pmndrs/zustand#2825] [Bug report] [answered]
### Bug Description

Hi, I was creating a little e commerce and I have a page where I display some info based on the global state with zustand. I get this info with an action within the store. This action returns an object which is the one I consume in my component as shown below: 

![image](https://github.com/user-attachments/assets/08c1c64b-39fc-4240-aee4-45cbc2e4d1b8)

If I try to consume this info from the other component I get different errors depending on how I do it. If I do it like this:

![image](https://github.com/user-attachments/assets/5d78d933-f92c-45f3-8bdb-5b1033455cc2)

I get this error: Error: Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.

On the other hand if I try to do it like this:

![image](https://github.com/user-attachments/assets/af92a856-38ec-4fc3-890e-b700de4a0574)

the error is not shown anymore but then the object gets unresponsive and the info doesn't update as the state changes. This only happens if I return an object. For single values it works just fine.

As you can see I'm generating the store the way zustand recommends in the docs when working with Next.js and provide it with the context api. However if I try to do it the regular way with create, the app work just as expected and all the info behaves responsive. 

![image](https://github.com/user-attachments/assets/952871ce-e36f-4e42-849c-381b6a4369ae)

I see the same behavior with both v4 and v5.  Since we use useStore when we integrate zustand with next, maybe it is related with this issue? https://github.com/pmndrs/zustand/discussions/2467#discussion-6469232

### Reproduction Link

NA

**Comment by dbritto-dev** [comment]
@samuelgomez4 not a bug but expected behavior, each time you call `getSummaryInfo` you are returning a different object that could cause some issues due zustand use `useSyncExternalStore` and require stable selector results in order to cache those results, without that you ended up on an infinite loop, one way to fix that would be using `useShallow` hook

**Comment by samuelgomez4** [comment]
Thanks for replying @dbritto-dev . I still wonder why the behaviour differs depending on how we create the store. What you mention is only valid for useStore?

**Reply by dbritto-dev** [reply]
@samuelgomez4 for all, we are talking about stable selector results

**Comment by samuelgomez4** [comment]
@dbritto-dev Makes sense, but then the fact that I get an infinite loop when I provide the store with the context vs the normal behavior I get when I do it the regular way might be related to the context API itself and not zustand?

**Reply by dbritto-dev** [reply]
@samuelgomez4 would you mind sharing us a minimal repro on stackblitz? I wouldn't say that is something related to context api that shouldn't be an issue. We have couple examples of that in our new guides on apis, hooks, and middleware sections

**Comment by samuelgomez4** [comment]
Hi @dbritto-dev, I was actually able to reproduce the error only in [version 5 of zustand](https://stackblitz.com/edit/vitejs-vite-wzwwfc?file=src%2FApp.tsx).  In [version 4](https://stackblitz.com/edit/vitejs-vite-tdcnuv?file=src%2FApp.tsx) works as I was expecting. I thought at some point I had the same error in v4 with the context but couldn't reproduce it again. I was reading the breaking changes of v5 and this one is one of them, sorry I missed that. Thanks!

**Reply by dbritto-dev** [reply, marked as answer]
@samuelgomez4 v4 hide some errors on v5 those errors are visible
